sapicore/engine/synapse/__init__.py

```
@@ -0,0 +1,330 @@
+            `prop`*`dst_ensemble.num_units` randomly selected destination neurons. "rand" sets a proportion `prop`
+            of the matrix elements to 1 with no balance constraints (unlike "prop").
+
+        prop:
+            Desired proportion a neurons in destination receiving a connection from a particular neuron
+            in the source ensemble. Rounded to an integer using np.round.
+
+            E.g., if prop = 0.2 and matrix_shape is (100, 5), 20 random row elements (destination neurons)
+            in the connectivity mask will be set to 1.
+
+        Warning
+        -------
+        Attempting to initialize one-to-one connectivity between ensembles of different sizes will necessarily
+        result in some unconnected neurons. The faux-diagonal will be set to 1s but there will be all 0s rows/cols,
+        depending on whether the source or destination ensemble is bigger.
+
+        """
+        match mode:
+            case "all":
+                self.connections = torch.ones_like(self.connections)
+
+            case "one":
+                self.connections = torch.eye(*self.connections.size(), device=self.device)
+
+            case "prop":
+                selection_size = int(np.round(prop * self.matrix_shape[0]))
+                torch.randperm(self.matrix_shape[0])
+
+                self.connections = torch.zeros_like(self.connections)
+
+                for i in range(self.matrix_shape[1]):
+                    self.connections[torch.randperm(self.matrix_shape[0])[:selection_size], i] = 1
+
+            case "rand":
+                num_enabled = int(np.round(prop * self.weights.numel()))
+                ids_enabled = np.random.choice(np.arange(self.weights.numel()), size=num_enabled, replace=False)
+
+                self.connections = torch.zeros_like(self.connections)
+                self.connections[np.unravel_index(ids_enabled, shape=self.matrix_shape)] = 1.0
+
+            case _:
+                return
```

This should raise an exception!?